File: GRM-Java/app/src/main/java/com/tartagliaeg/grmjava/domain/utils/DataSource.java
package com.tartagliaeg.grmjava.domain.utils;


import io.reactivex.Single;
import io.reactivex.SingleSource;
import io.reactivex.functions.Function;

public class DataSource<T> {
  private final String mSource;
  private final T mData;

  private static final String NETWORK = "NETWORK";
  private static final String MEMORY = "MEMORY";
  private static final String DATABASE = "DATABASE";
  private static final String NONE = "NONE";

  private DataSource(String source, T data) {
    this.mSource = source;
    this.mData = data;
  }

  public static <NewType> DataSource<NewType> empty(NewType emptyRepresentation) {
    return new DataSource<>(NONE, emptyRepresentation);
  }

  public static <NewType> DataSource<NewType> empty() {
    return new DataSource<>(NONE, null);
  }

  public static <NewType> DataSource<NewType> from(DataSource source, NewType newType) {
    return new DataSource<>(source.mSource, newType);
  }

  public static <NewType> Single<DataSource<NewType>> memory(Single<NewType> data) {
    return data.flatMap(new Function<NewType, SingleSource<? extends DataSource<NewType>>>() {
      @Override
      public SingleSource<? extends DataSource<NewType>> apply(NewType newType) throws Exception {
        return Single.just(memory(newType));
      }
    });
  }

  public static <NewType> Single<DataSource<NewType>> database(Single<NewType> data) {
    return data.flatMap(new Function<NewType, SingleSource<? extends DataSource<NewType>>>() {
      @Override
      public SingleSource<? extends DataSource<NewType>> apply(NewType newType) throws Exception {
        return Single.just(database(newType));
      }
    });
  }

  public static <NewType> Single<DataSource<NewType>> network(Single<NewType> data) {
    return data.flatMap(new Function<NewType, SingleSource<? extends DataSource<NewType>>>() {
      @Override
      public SingleSource<? extends DataSource<NewType>> apply(NewType newType) throws Exception {
        return Single.just(network(newType));
      }
    });
  }


  public static <NewType> DataSource<NewType> memory(NewType data) {
    return new DataSource<>(MEMORY, data);
  }

  public static <NewType> DataSource<NewType> database(NewType data) {
    return new DataSource<>(DATABASE, data);
  }

  public static <NewType> DataSource<NewType> network(NewType data) {
    return new DataSource<>(NETWORK, data);
  }

  public boolean isEmpty() { return NONE.equals(mSource); }

  public boolean isMemory() { return MEMORY.equals(mSource); }

  public boolean isDatabase() { return DATABASE.equals(mSource); }

  public boolean isNetwork() { return NETWORK.equals(mSource); }

  public T getData() {
    return mData;
  }

}

File: GRM-Swift/GRM-Swift/Domain/DataSource.swift
//
//  DataSource.swift
//  GRM-Swift
//
//

import Foundation
import RxSwift

@objc enum DataSourceOrigin: Int {
    case NETWORK
    case MEMORY
    case DATABASE
    case NONE
}

class DataSource<T> {
    private let _source: DataSourceOrigin;
    private let _data: T?;
    
    var data: T? { get { return self._data } }
    
    init(source: DataSourceOrigin, data: T?) {
        self._source = source;
        self._data = data;
    }
    
    static func empty<NewType>(_ emptyRepresentation: NewType) -> DataSource<NewType> {
        return DataSource<NewType>.init(source: DataSourceOrigin.NONE, data: emptyRepresentation);
    }
    
    static func empty<NewType>() -> DataSource<NewType> {
        return DataSource<NewType>.init(source: DataSourceOrigin.NONE, data: nil);
    }

    static func from<NewType>(_ source: DataSource,_  newType: NewType) -> DataSource<NewType> {
        return DataSource<NewType>.init(source: source._source, data: newType);
    }

    static func memory<NewType>(_ data: NewType) -> DataSource<NewType> {
        return DataSource<NewType>.init(source: DataSourceOrigin.MEMORY, data: data);
    }

    static func database<NewType>(_ data: NewType) -> DataSource<NewType> {
        return DataSource<NewType>.init(source: DataSourceOrigin.DATABASE, data: data);
    }

    static func network<NewType>(_ data: NewType) -> DataSource<NewType> {
        return DataSource<NewType>.init(source: DataSourceOrigin.NETWORK, data: data);
    }

    static func memory<NewType>(_ stream: Single<NewType>) -> Single<DataSource<NewType>> {
        return stream.flatMap { Single.just(self.memory($0));  }
    }

    static func database<NewType>(_ stream: Single<NewType>) -> Single<DataSource<NewType>> {
        return stream.flatMap { Single.just(self.database($0));  }
    }

    static func network<NewType>(_ stream: Single<NewType>) -> Single<DataSource<NewType>> {
        return stream.flatMap { Single.just(self.network($0));  }
    }

    static func memory<NewType>(_ stream: Observable<NewType>) -> Observable<DataSource<NewType>> {
        return stream.flatMap { Observable.just(self.memory($0));  }
    }
    
    static func database<NewType>(_ stream: Observable<NewType>) -> Observable<DataSource<NewType>> {
        return stream.flatMap { Observable.just(self.database($0));  }
    }
    
    static func network<NewType>(_ stream: Observable<NewType>) -> Observable<DataSource<NewType>> {
        return stream.flatMap { Observable.just(self.network($0));  }
    }

    
    func isEmpty() -> Bool { return DataSourceOrigin.NONE == _source; }
    func isMemory() -> Bool { return DataSourceOrigin.MEMORY == _source; }
    func isDatabase() -> Bool { return DataSourceOrigin.DATABASE == _source; }
    func isNetwork() -> Bool { return DataSourceOrigin.NETWORK == _source; }
    
}
